pyresample/geometry.py

```diff
@@ -713,43 +715,301 @@ def copy(self):
         return np.dstack((x, y, z))
 
     def aggregate(self, **dims):
-        """Aggregate the current swath definition by averaging.
-
-        For example, averaging over 2x2 windows:
-        `sd.aggregate(x=2, y=2)`
+        """Return an aggregated version of the area."""
+        warnings.warn("'aggregate' is deprecated, use 'downsample' instead.", PendingDeprecationWarning)
+        return self.downsample(x=dims.get('x', 1), y=dims.get('y', 1))
+
+    def downsample(self, x=1, y=1, **kwargs):
+        """Downsample the SwathDefinition along x (columns) and y (lines) dimensions.
+
+        Builds upon xarray.DataArray.coarsen averaging function.
+        To downsample of a factor of 2, call swath_def.downsample(x=2, y=2)
+        swath_def.downsample(x=1, y=1) simply returns the current swath_def.
+        By default, it raise a ValueError if the dimension size is not a multiple of the window size.
+        This can be changed by passing boundary="trim" or boundary="pad", but behaviour within pyresample is undefined.
+        See https://xarray.pydata.org/en/stable/generated/xarray.DataArray.coarsen.html for further details.
         """
         import dask.array as da
-        import pyproj
+        import xarray as xr
+
+        # Check input validity
+        x = int(x)
+        y = int(y)
+        if x < 1 or y < 1:
+            raise ValueError('SwathDefinition.downsample expects (integer) aggregation factors >=1 .')
+
+        # Return SwathDefinition if nothing to downsample
+        if x == 1 and y == 1:
+            return self
 
+        # Define geodetic and geocentric projection
         geocent = pyproj.Proj(proj='geocent')
         latlong = pyproj.Proj(proj='latlong')
+
+        # Get xr.DataArray with dask array
+        # - If input lats/lons are xr.DataArray, the specified dims ['x','y'] are ignored
+        src_lons, src_lons_format = _convert_2D_array(self.lons, to='DataArray_Dask', dims=['y', 'x'])
+        src_lats, src_lats_format = _convert_2D_array(self.lats, to='DataArray_Dask', dims=['y', 'x'])
+
+        # Conversion to Geocentric Cartesian (x,y,z) CRS
         res = da.map_blocks(self._do_transform, latlong, geocent,
-                            self.lons.data, self.lats.data,
-                            da.zeros_like(self.lons.data), new_axis=[2],
-                            chunks=(self.lons.chunks[0], self.lons.chunks[1], 3))
-        res = DataArray(res, dims=['y', 'x', 'coord'], coords=self.lons.coords)
-        res = res.coarsen(**dims).mean()
+                            src_lons.data,
+                            src_lats.data,
+                            da.zeros_like(src_lons),  # altitude
+                            new_axis=[2],
+                            chunks=(src_lons.chunks[0], src_lons.chunks[1], 3))
+        res = xr.DataArray(res, dims=['y', 'x', 'xyz'], coords=src_lons.coords)
+
+        # Aggregating
+        res = res.coarsen(x=x, y=y, **kwargs).mean()
```

Comment:
This should probably wait for a separate PR, but it'd be nice if `mean` was a keyword argument that the user could control.

Reply:
I can do also this.
There is a small set of possible functions that can be called from an [xarray.core.rolling.DataArrayCoarsen](https://xarray.pydata.org/en/stable/generated/xarray.core.rolling.DataArrayCoarsen.html#xarray.core.rolling.DataArrayCoarsen) object. I can add the default argument` fun="mean"`  and then do:
```
res = res.coarsen(x=x, y=y, **kwargs)
func = getattr(res, fun')
res = func(res)
```
Should I check for valid `fun` values in `set("min", "max", "sum", "std", "var", "count", "mean", "median") `? Or defer this to the xarray function?

Reply:
I'd say add an explicit keyword argument in the method signature for `coarsen_func` or something like that and document in the docstring to look at the xarray docs (link to the exact page if possible, intersphinx may allow you to do something like:

```
See the :doc:`Xarray coarsen documentation <xarray:coarsen-docs-page>` for more information.
```

And I think you could leave the error handling up to the `getattr`. You could shorten this to:

```
coarsen_res = res.coarsen(...)
res = getattr(coarsen_res, coarsen_func)()
```

This assumes the coarsen function doesn't take additional kwargs.